### PR TITLE
 *: Add initial version of a distributed monotonic id generator

### PR DESCRIPTION
**Pull request by mxinden**

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+"checksum thread_local 0.3.6 (registry+https://github.com/rust-lang/crates.io-index)" = "c6b53e329000edc2b34dbe8545fd20e55a333362d0a321909685a19bd28c3f1b"
+"checksum ucd-util 0.1.3 (registry+https://github.com/rust-lang/crates.io-index)" = "535c204ee4d8434478593480b8f86ab45ec9aae0e83c568ca81abf0fd0e88f86"
+"checksum unicode-xid 0.1.0 (registry+https://github.com/rust-lang/crates.io-index)" = "fc72304796d0818e357ead4e000d19c9c174ab23dc11093ac919054d20a6a7fc"
+"checksum utf8-ranges 1.0.2 (registry+https://github.com/rust-lang/crates.io-index)" = "796f7e48bef87609f7ade7e06495a87d5cd06c7866e6a5cbfceffc558a243737"
+"checksum winapi 0.3.7 (registry+https://github.com/rust-lang/crates.io-index)" = "f10e386af2b13e47c89e7236a7a14a086791a2b88ebad6df9bf42040195cf770"
+"checksum winapi-i686-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" = "ac3b87c63620426dd9b991e5ce0329eff545bccbbb34f3be09ff6fb6ab51b7b6"
+"checksum winapi-x86_64-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"
```
Comment (spacejam) [inline Cargo.lock:413]: Cargo.lock files are good to check in for anything that is meant to be distributed as a standalone binary. For things that are meant to be used as a dependency, no lock should be checked in. Just saying this in case the idiom is interesting for you to hear.

File: Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+authors = ["Max Leonard Inden <IndenML@gmail.com>"]
+edition = "2018"
+
+[dependencies]
+queues = "1.0.0"
+rand  = "*"
+
+[dev-dependencies]
+quickcheck = "0.8"
+quickcheck_macros = "0.8"
```
Comment (spacejam) [inline Cargo.toml:13]: In some cases it's better to use the proptest crate than quickcheck, as it will create a file that contains previously failing seeds for the generator, so that it automatically replays failing tests (unless you change the generator). I use both, and tend to use quickcheck for model-like tests like simulators, since the Arbitrary instance is going to change a lot over time. This kind of engineering is like a conversation between your model and the implementation to fix both of them until they line up.

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
+use std::collections::HashMap;
+
+pub type ID = u64;
```
Comment (spacejam) [inline src/lib.rs]: Usually for multi-letter abbreviations in rust we only capitalize the first letter. This felt weird for me for a while, but it's the idiom and now I accept it. So, Http not HTTP, Id, not ID etc...

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+pub type Addr = String;
+pub type Timestamp = usize;
+
+pub trait Receiver {
+    fn receive(&mut self, m: Msg, addr: Addr) -> Vec<(Msg, Addr)>;
```
Comment (spacejam) [inline src/lib.rs]: an `_: Msg` might be better than a single-letter name here, since you're not trying to be descriptive with the name

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        match m {
+            Msg::StartRequest => {
+                panic!();
+            }
+            Msg::Request(id) => {
+                if self.highest_id_seen < id {
+                    self.highest_id_seen = id;
+                    return vec![(Msg::Yes(id), from)];
+                }
+                return vec![(Msg::No(id), from)];
```
Comment (spacejam) [inline src/lib.rs]:
```
                if self.highest_id_seen < id {
                    self.highest_id_seen = id;
                    vec![(Msg::Yes(id), from)]
                } else {
                    vec![(Msg::No(id), from)]
                }
```

kind of on the fence about which one I prefer. generally avoid rightword pressure (as I added) but `else` maybe feels lighter than 2 explicit returns? This is seriously splitting hairs but I think I might avoid the 2 returns so close to each other.

File: tests/integration_test.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
+extern crate id_generator;
+extern crate rand;
+
+#[cfg(test)]
+extern crate quickcheck;
+#[cfg(test)]
+#[macro_use(quickcheck)]
+extern crate quickcheck_macros;
```
Comment (spacejam) [inline tests/integration_test.rs]: extern crate and macro_use is no longer necessary with edition 2018

Comment (spacejam) [reply tests/integration_test.rs]: macros need to be imported like a normal function though

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+impl Receiver for Client {
+    fn receive(&mut self, m: Msg, _addr: Addr) -> Vec<(Msg, Addr)> {
+        let to_all_servers = |servers: &Vec<Addr>, msg: Msg| -> Vec<(Msg, Addr)> {
+            servers
+                .iter()
+                .map(|server| (msg.clone(), server.clone()))
+                .collect()
+        };
```
Comment (spacejam) [inline src/lib.rs:149]: `to_all_servers` might feel a little more natural as a normal self receiver method, since it always is passed self.servers. this could also be a nested fn. generally, use a closure when you close over some external state, or when it keeps the usage of combinators like `map` simpler. otherwise a (nested) fn is sometimes better, since it's more restrictive and therefore less likely to surprise you or be annoying to refactor.

File: src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+        let get_responses_with_default = |id: ID| -> (usize, usize) {
+            match self.responses.get(&id) {
+                Some((y, n)) => (y.clone(), n.clone()),
+                None => (0, 0),
+            }
```
Comment (spacejam) [inline src/lib.rs]: `self.responses.get(&id).unwrap_or((0, 0))`

File: src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        };
+
+        match m {
+            Msg::StartRequest => {
+                self.highest_id_seen = self.highest_id_seen + 1;
```
Comment (spacejam) [inline src/lib.rs]: += 1

File: src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+                return to_all_servers(&self.servers, Msg::Request(self.highest_id_seen));
```
Comment (spacejam) [inline src/lib.rs]: can skip all explicit return statements since the match block is the last part of the method body

File: src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+            }
+            Msg::Yes(id) => {
+                let (yes, no) = get_responses_with_default(id);
+
+                self.responses.insert(id, (yes + 1, no));
+
+                // '==' not '>=' to prevent double adding.
+                if yes == self.servers.len() / 2 {
+                    self.claimed_ids.push(id.clone());
```
Comment (spacejam) [inline src/lib.rs]: `*id` instead of clone. This is because u64 implements Copy, which means it's just a bunch of dumb bytes (no pointers) with no Drop impl and therefore safe to make copies of automatically.

File: tests/integration_test.rs [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+use quickcheck::{quickcheck, TestResult};
+
+mod simulator;
+
+#[cfg(test)]
+mod tests {
```
Comment (spacejam) [inline tests/integration_test.rs]: since this exists in a module in the `tests/` directory it's already conditionally compiled, so you can de-nest all of this without the submodule.

File: tests/integration_test.rs [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    use super::*;
+
+    #[test]
+    fn basic_run() {
+        let mut simulator = simulator::Simulator::new(2, 2, 2, 2);
+
+        match simulator.run() {
+            Ok(_) => {}
+            Err(e) => panic!(e),
+        };
```
Comment (spacejam) [inline tests/integration_test.rs]: simulator.run().unwrap();

File: tests/integration_test.rs [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+        let validation = simulator.validate_run();
+
+        match validation {
+            Ok(_) => {}
+            Err(e) => panic!(e),
+        }
```
Comment (spacejam) [inline tests/integration_test.rs]: simulator.validate_run().unwrap();

File: tests/integration_test.rs [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    #[test]
+    fn one_client_two_servers() {
+        let mut simulator = simulator::Simulator::new(1, 2, 1, 2);
+
+        match simulator.run() {
+            Ok(_) => {}
+            Err(e) => panic!(e),
+        };
```
Comment (spacejam) [inline tests/integration_test.rs]: unwrap

File: tests/integration_test.rs [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+        let validation = simulator.validate_run();
+
+        match validation {
+            Ok(_) => {}
+            Err(e) => panic!(e),
+        }
+    }
```
Comment (spacejam) [inline tests/integration_test.rs]: unwrap

File: tests/integration_test.rs [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+        if clients > 100 || servers > 100 || ids > 1000 || delay > 1000 {
+            return TestResult::discard();
+        }
+
+        let mut simulator = simulator::Simulator::new(clients, servers, ids, delay);
+
+        match simulator.run() {
+            Ok(_) => {}
+            Err(e) => return TestResult::error(e),
+        };
```
Comment (spacejam) [inline tests/integration_test.rs:76]: `simulator.run().map_err(TestResult::error)?;`

enum variants are functions :]

File: tests/simulator.rs [hunk 18]
```diff
@@ -0,0 +1,201 @@
+extern crate id_generator;
+extern crate queues;
```
Comment (spacejam) [inline tests/simulator.rs]: no need for extern crate

File: tests/simulator.rs [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    pub fn run(&mut self) -> Result<(), String> {
+        loop {
+            if (self.goal_per_client * self.servers.len() * self.clients.len() * 100) < self.time {
+                return Err("too many iterations".to_string());
+            }
+
+            if self.in_flight.len() == 0 {
+                break;
```
Comment (spacejam) [inline tests/simulator.rs]: I'd just return Ok(()) here since it's the only break

Comment (spacejam) [reply tests/simulator.rs]: or replace the loop with a while statement

File: tests/simulator.rs [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+                Some(server) => (server.get_addr().clone(), server.receive(e.msg, e.from)),
+                None => panic!(),
+            },
+            Msg::Yes(_) => match self.clients.get_mut(&e.to) {
+                Some(client) => (client.get_addr().clone(), client.receive(e.msg, e.from)),
+                None => panic!(),
+            },
+            Msg::No(_) => match self.clients.get_mut(&e.to) {
+                Some(client) => (client.get_addr().clone(), client.receive(e.msg, e.from)),
+                None => panic!(),
```
Comment (spacejam) [inline tests/simulator.rs]: all the matches where None panics can be replaced by `self.clients.get_mut(&e.to).unwrap()`

Comment (spacejam) [reply tests/simulator.rs]: this can happen before the match of msg

File: tests/simulator.rs [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+            for j in (i + 1)..self.clients.len() {
+                let b = clients[j];
+
+                for a_id in a.claimed_ids.iter() {
+                    for b_id in b.claimed_ids.iter() {
+                        if a_id == b_id {
+                            return Err(format!(
+                                "both client {} and {} claimed id {}",
+                                a.addr, b.addr, a_id
+                            ));
```
Comment (spacejam) [inline tests/simulator.rs]: can maybe just insert claimed ID's into a map and assert that insert returns None (insert returns the previous element, or None if it's the first insert)

File: tests/simulator.rs [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+    use super::*;
+
+    #[test]
+    fn simulator_increases_timestamp() {
+        let mut sim = Simulator::new(2, 3, 1, 2);
+
+        match sim.run() {
+            Ok(_) => {}
+            Err(e) => panic!(e),
+        };
```
Comment (spacejam) [inline tests/simulator.rs:215]: run().unwrap()

**Comment by spacejam** [comment]
Overall this really looks great! Nice work :]

**Comment by spacejam** [comment]
Also:
```
rustup component add clippy
cargo clippy
```

This might catch some things I missed, and is generally really good

File: Cargo.toml (hunk 2 above)
Comment (lucab) [reply Cargo.toml:13]: If you have some time and care about this, I'd be happy if you could expand on the proptest-vs-quickcheck rationale somewhere more public. I've been blindly using proptest so far.

File: src/lib.rs (hunk 9 above)
Comment (lucab) [reply src/lib.rs]: `.saturating_add(1)`? `.wrapping_add(1)`? What's the semantic when this reaches its maximum value?

File: Cargo.toml (hunk 2 above)
Comment (spacejam) [reply Cargo.toml:13]: I think it's a good idea to use proptest as a default, since it gives you so much more out of the box. The reason I still use quickcheck basically boils down to the fact that it's a much simpler library, and fewer moving parts makes it a better foundation when I need to implement a lot of complexity on my own anyway without leaning much into the complexity of proptest. I'll use quickcheck when the value of proptest isn't as high: very complex generators that are likely to change and result in the proptest persisted seed file no longer correlating to an "interesting" behavior. I'll basically just manually create "interesting" tests with quickcheck-backed complex generators to guarantee the regression test doesn't change when the generator does. 

[complex property example where I like quickcheck](https://github.com/spacejam/sled/blob/5ce7b818991598de6d645a779156b2477fb373dd/tests/src/tree.rs#L202)
[big file of tests, manually copy + pasted + often postmortem commented](https://github.com/spacejam/sled/blob/5ce7b818991598de6d645a779156b2477fb373dd/tests/tests/test_tree.rs#L778)

For sled, I'll add more operations over time that I want to test. I'll just throw the new operation into the Arbitrary instance, which completely changes what gets generated based on a seed, but because I manually copy interesting failures I keep my interesting regressions.

I use proptest by default because it's so much more fully-featured, and it's pretty much just these complex model tests that I use quickcheck for. When writing my [model](https://docs.rs/mode) crate I used proptest though because I could use the rich set of default generators as a building block, and the regression file support is really nice even though maybe sometimes people change their models.

File: src/lib.rs (hunk 9 above)
Comment (mxinden) [reply src/lib.rs]: @lucab is this a general question on specific to this use case? I guess even though `Id` is a `u64` one would need to take care of the _wrap-around-case_ at some point.

Comment (lucab) [reply src/lib.rs]: Both. When this reaches `u64::MAX`, it gets different behavior in release vs. debug mode. If it is an interesting corner-case, you'd want to make your intended behavior explicit.

File: tests/integration_test.rs (hunk 17 above)
Comment (mxinden) [reply tests/integration_test.rs:76]: Yes, but the function does not return an option or a result, thus I can't use the `?` operator. How can I make this more idiomatic?

```
error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
  --> tests/integration_test.rs:73:9
   |
73 |         simulator.run().map_err(TestResult::error)?;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `quickcheck::tester::TestResult`
   |
   = help: the trait `std::ops::Try` is not implemented for `quickcheck::tester::TestResult`
   = note: required by `std::ops::Try::from_error`

error: aborting due to previous error
```

**Comment by mxinden** [comment]
Thanks a bunch for all the comments @spacejam and @lucab. Quickcheck hasn't found any further regressions for 30 minutes. I guess next up would be to drop messages in the simulator. So far I am only delaying them.
